### PR TITLE
Holes!

### PR DESCRIPTION
**Pull request by FlattestGuitar**
- Step on one while walking - nothing!
- Step on one while running - ouch!
- Step on one while running and there are spikes inside - oh god!

Possibly we can add something like covering these up with leaves later on, but I've got no sprites and this stuff as is already caused me enough sweat in Photoshop.

Despite the hacky implementation, the overlay objects don't actually show up in-game, so that's not an issue.

### CODE REVIEW

File: code/game/objects/structures/pitfall.dm [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	has_spikes = TRUE
+	icon_state = "hole_spikes"
+
+/obj/hole_overlay_fuck_byond
+	mouse_opacity = 0
+	layer = 1.8
+	icon = 'icons/obj/hole.dmi'
+	icon_state = "hole_mask"
+
+/obj/hole_overlay_fuck_byond/Initialize(location, var/obj/structure/hole/parent_hole)
```
Comment (Citinited) [inline code/game/objects/structures/pitfall.dm]: 👌

File: code/game/objects/structures/pitfall.dm [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	var/obj/hole_overlay_fuck_byond/overlay
+	var/turf/base_turf
+	var/has_spikes = FALSE
+
+/obj/structure/hole/spikes
+	desc = "There are spikes down there. Looks like a very intricate barbecue setup."
+	has_spikes = TRUE
+	icon_state = "hole_spikes"
+
+/obj/hole_overlay_fuck_byond
```
Comment (Citinited) [inline code/game/objects/structures/pitfall.dm]: nice overlay name 👌

File: code/game/objects/structures/pitfall.dm [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	for(var/mob/living/L in loc)
+		L.layer = 1.5
+		animate(L, pixel_y = L.pixel_y - 6, time = 2)
+
+/obj/structure/hole/Destroy()
+	base_turf.layer = initial(base_turf.layer)
+	QDEL_NULL(overlay)
+
+/obj/structure/hole/Crossed(atom/movable/AM)
+	if(!istype(AM, /mob/living))
```
Comment (Citinited) [inline code/game/objects/structures/pitfall.dm]: Won't this stop people from being able to throw things across the hole / fire bullets / other?

File: code/game/objects/structures/pitfall.dm [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+		if(has_spikes)
+			to_chat(L, "<span class='userdanger'>You trip and fall into [src], impaling yourself on the sharp spikes!\nIt hurts!</span>")
+			L.Stun(100)
+			L.Knockdown(100)
+			L.adjustBruteLoss(40)
+		else
+			to_chat(L, "<span class='userdanger'>You trip and fall into [src]!</span>")
+			L.Stun(50)
+			L.Knockdown(50)
+			playsound(loc, 'sound/weapons/slice.ogg', 50, 1, -1)
```
Comment (Citinited) [inline code/game/objects/structures/pitfall.dm]: This should be in the `has_spikes` bit, not here. Maybe a fleshy thud sound would work here

File: code/game/objects/structures/pitfall.dm (hunk 2 above)
Comment (Citinited) [inline code/game/objects/structures/pitfall.dm]: nice overlay name 👌

File: code/game/objects/structures/pitfall.dm [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+/obj/structure/hole/Uncrossed(atom/movable/AM)
+	if(!istype(AM, /mob/living))
+		return
+	var/mob/living/L = AM
+
+	L.layer = initial(L.layer)
+	animate(L, pixel_y = L.pixel_y + 6, time = 2)
```
Comment (Citinited) [inline code/game/objects/structures/pitfall.dm]: Maybe have some text like `to_chat(L, "<span class = 'notice'>You climb out of [src].</span>)"`?

File: code/game/objects/structures/pitfall.dm [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	base_turf.layer = initial(base_turf.layer)
+	QDEL_NULL(overlay)
+
+/obj/structure/hole/Crossed(atom/movable/AM)
+	if(!istype(AM, /mob/living))
+		return
+	var/mob/living/L = AM
+
+	L.layer = 1.5
+	animate(L, pixel_y = L.pixel_y - 6, time = 2)
```
Comment (Citinited) [inline code/game/objects/structures/pitfall.dm]: If they fall into the pit, this animation shouldn't happen, it should be abrupt.

**Comment by Citinited** [comment]
It would be nice if two pits next to each other formed a trench, maybe four in a square formed a larger hole? That would open up a whole bunch of possibilities (but would also require reworking this, so IDK if you want to save it for another PR).

File: code/game/objects/structures/pitfall.dm (hunk 5 above)
Comment (SamHPurp) [inline code/game/objects/structures/pitfall.dm]: probably have it take some time too

**Comment by FlattestGuitar** [comment]
> IDK if you want to save it for another PR
Yes, I do.

**Pull request merged**